### PR TITLE
Add AnyPrecisionAdamW optimizer

### PR DESCRIPTION
**Pull request by atturaioe**
# What does this PR do?
Add `AnyPrecisionAdamW` optimizer from `torchdistx`

Fixes # (issue)
#18827

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@stas00

### CODE REVIEW

**Comment by atturaioe** [comment]
Hello, @stas00, I want to ask you whether should I add `anyprecision_adamw` specific arguments to the `trainings_args.py` or use the default ones in `trainer.py`. I'll be working on tests.

**Comment by stas00** [comment]
I'd say let's add a generic `--optim-args` optional arg which can then supply options to any future optimizer - i.e. it'd pair with `--optim`.

I'm trying to remember if we already have the plumbing for parsing in place - I think the `--debug` flag has it.

But something like `--optim-args "key1:val1; key2:val2; ..."`

**Comment by stas00** [comment]
@atturaioe, this is just another variation - perhaps `--optim-args` can support just the exact syntax as python function sig?

```
--optim anyprecision_adamw --optim-args "use_kahan_summation=true, momentum_dtype=torch.bf1oat16; ..."
```

so `,` separator and perhaps writing out the dtypes exactly as they are in python and converting them on the fly to an actual class name.

Perhaps it'd be easier to mimick the signature. Not sure. Let's see what you think is better.

**Comment by atturaioe** [comment]
Yeah! But should I parse the `--optim-args` into `dict` or something like that right in the `trainer.get_optimizer_cls_and_kwargs`?

**Comment by stas00** [comment]
Yes, that's exactly right:

https://github.com/huggingface/transformers/blob/d842f2d5b9bd4e361644c332bf9dc7f9b064f581/src/transformers/trainer.py#L1094

**Comment by atturaioe** [comment]
Is it any good? 
I didn't quite understand about converting dtypes on the fly(using `eval`?).

**Comment by stas00** [comment]
eval would be unsafe. here is a quick proof of concept:

```
python -c "import torch; x = 'torch.float16'; print(getattr(torch, x.split('.')[1]))"
```

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -113,6 +113,7 @@ class OptimizerNames(ExplicitEnum):
     ADAMW_APEX_FUSED = "adamw_apex_fused"
     ADAFACTOR = "adafactor"
     ADAMW_BNB = "adamw_bnb_8bit"
+    ANYPRECISION_ADAMW = "anyprecision_adamw"
```
Comment (stas00) [inline src/transformers/training_args.py:116]: Perhaps let's just call it "adamw_anyprecision" to match all the other `/^adam_/ optims?

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1142,6 +1152,24 @@ def get_optimizer_cls_and_kwargs(args: TrainingArguments) -> Tuple[Any, Any]:
...
+                optimizer_cls = AnyPrecisionAdamW
+                optimizer_kwargs.update(adam_kwargs)
+                optimizer_kwargs.update(
+                    {
+                        "use_kahan_summation": strtobool(optim_args.get("use_kahan_summation", "False")),
+                        "momentum_dtype": getattr(torch, optim_args.get("momentum_dtype", "float32")),
+                        "variance_dtype": getattr(torch, optim_args.get("variance_dtype", "bfloat16")),
+                        "compensation_buffer_dtype": getattr(
+                            torch, optim_args.get("compensation_buffer_dtype", "bfloat16")
+                        ),
```
Comment (stas00) [inline src/transformers/trainer.py:1168]: Let's discuss which defaults would be the most beneficial so that it shines out of the box.

Probably `use_kahan_summation=True`, no?

**Review by stas00 — COMMENTED** [review]
So far excellent work, @atturaioe 

Let's figure out a few small bits and then add tests and I think this should be great.

File: src/transformers/trainer.py (hunk 2 above)
Comment (atturaioe) [reply src/transformers/trainer.py:1168]: You're right, I just made them as the default in `AnyPrecisionOptimizer`.

Comment (lessw2020) [reply src/transformers/trainer.py:1168]: Hi @stas00  and @atturaioe,
The best/most impressive results so far are definitely running in pure BF16 (so momentum and var set to torch.bfloat16) and use_kahan_summation=True. 

The caveat here is that the model itself needs to be independently set outside of the optimizer to BFloat16 to match (i.e. model.to(torch.bfloat16). 
Are you able to tie the model running in bf16 directly so the user does not have to do that part?

Comment (stas00) [reply src/transformers/trainer.py:1168]: unfortunately not for training, but yes for eval, here are the different available  automatic dtype knobs -

https://github.com/huggingface/transformers/blob/4157e3cd7e2bb5a7be6dc065a3e20c49cc1300ab/src/transformers/training_args.py#L271-L291

 `--bf16` here means mixed precision.

-----

so practically what will happen to master weights then? In mixed precision you'd want to have master weights in fp32 - if we switch to `model.to(torch.bfloat16)` this can be quite lossy - as there will be no kahan summation here. it's possible I'm missing some important nuance here.

----

Specifically to what we have right now - perhaps we can for now mark this optimizer as experimental and tune it up as users start to use it?

Comment (lessw2020) [reply src/transformers/trainer.py:1168]: Thanks for this info @stas00 
1 - For mixed precision - you could either 
a - run with the current defaults (M=fp32, Var = BF16, Kahan = False) and that would provide the memory and speed improvements from the Variance in BF16.  That works nicely, and you can make that all work 'automatically' per above control options.
b - you could also go all BF16 (M=BF16, Var = BF16, Kahan = False) because you will still get high precision weight updates with the master weights being in fp32.  This is not as well tested though, but is something we are going to enable in FSDP soon by moving the working weight gradients to BF16, meaning you only have FP32 weights, nothing else. 

To your question - having the weights in BF16 (via model.to) will only work if Kahan summation is active.  If you don't run it with Kahan, then you are exactly right, you will hit weight stagnation and it will not be performant.  
The addition of Kahan is what makes it all work nicely.  

Re: mark as experimental and tune as users run with it - that sounds like a great idea.    I would just go ahead and use the current defaults then (M=FP32, Var = BF16, Kahan = False) as it's plug and play into FP32 or BF16 mixed precision.
I'm working on a video tutorial now actually for this optimizer.  Maybe we can add to the video once this PR is in, and show people how to run it with the manual change of model.to() and setting the defaults directly to get people comfortable with running in pure BF16.

Comment (stas00) [reply src/transformers/trainer.py:1168]: That's perfect, Less. Let's do what you suggest.

Let me copy your last comments out of this conversation to a normal comment, so that when this gets resolved it won't disappear. as we will want to eventually cover all these use-cases out of the box.

@atturaioe, so let's keep your defaults for now. and then tweak those in the future.

**Comment by stas00** [comment]
Just pasting @lessw2020's comment from https://github.com/huggingface/transformers/pull/18961#discussion_r970160808 so that it doesn't get hidden by github once resolved and we will want to revisit this down the road and support other configs:

> 1 - For mixed precision - you could either
> a - run with the current defaults (M=fp32, Var = BF16, Kahan = False) and that would provide the memory and speed improvements from the Variance in BF16. That works nicely, and you can make that all work 'automatically' per above control options.
> b - you could also go all BF16 (M=BF16, Var = BF16, Kahan = False) because you will still get high precision weight updates with the master weights being in fp32. This is not as well tested though, but is something we are going to enable in FSDP soon by moving the working weight gradients to BF16, meaning you only have FP32 weights, nothing else.
> 
> To your question - having the weights in BF16 (via model.to) will only work if Kahan summation is active. If you don't run it with Kahan, then you are exactly right, you will hit weight stagnation and it will not be performant.
> The addition of Kahan is what makes it all work nicely.
> 
> Re: mark as experimental and tune as users run with it - that sounds like a great idea. I would just go ahead and use the current defaults then (M=FP32, Var = BF16, Kahan = False) as it's plug and play into FP32 or BF16 mixed precision.
> I'm working on a video tutorial now actually for this optimizer. Maybe we can add to the video once this PR is in, and show people how to run it with the manual change of model.to() and setting the defaults directly to get people comfortable with running in pure BF16.

**Comment by atturaioe** [comment]
Checks will fail https://github.com/pytorch/torchdistx/issues/68

**Comment by atturaioe** [comment]
Hi @stas00, hope you had great time!
So the problem here that the `momentum_dtype` and `variance_dtype` set to different `dtypes` (`float32/bfloat16`) don't get casted dynamically, unless they're both of the same `dtype`. But of course I can set them both to same `dtype`.
Please correct me if I misunderstood something here.

**Comment by stas00** [comment]
Let's perhaps start with using the same dtype only and deal with that unusual case down the road should someone actually want to use it?

**Comment by atturaioe** [comment]
So I changed the default params to the `float32` since we have 2 options here for them to be the same dtype:
1 - `float32`
2 - `bfloat16` - but it won't pass tests since we have to move the `model.to(torch.bfloat16) `while running tests

**Comment by stas00** [comment]
ok, so as it has been awhile since this was created please rebase to master and flip the Draft mode to ready and we can then ask Sylvain to have a last look and merge.

**Review by stas00 — APPROVED** [review]
LGTM.

Let's have @sgugger have a last look and we are good to merge.

thank you very much for working on it, @atturaioe!

File: src/transformers/training_args.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/training_args.py:116]: This does not seem to have been addressed.

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -401,7 +402,9 @@ class TrainingArguments:
 
             The options should be separated by whitespaces.
         optim (`str` or [`training_args.OptimizerNames`], *optional*, defaults to `"adamw_hf"`):
-            The optimizer to use: adamw_hf, adamw_torch, adamw_apex_fused, or adafactor.
+            The optimizer to use: adamw_hf, adamw_torch, adamw_apex_fused, anyprecision_adamw or adafactor.
+        optim_args (`str`, *optional*):
+            Optional arguments to supply to optimizer.
```
Comment (sgugger) [inline src/transformers/training_args.py:407]: Mention here that this is only used for anyprecision_adam

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. There seems to be an unresolved conversation about naming?

**Review by sgugger — APPROVED** [review]
Thanks for iterating!
